### PR TITLE
fix(DTFS2-6293): Cookie security

### PR DESCRIPTION
**Pull request by abhi-markan**
## Introduction
Cookies being used are not using appropriate cookie prefix. This has been flagged in tech-debt.

## Resolution
* Introduction of `__Host-` cookie prefix. This prefix is only valid with `secure` flag and with path set to `\` while no domain has been specified.
* Renamed session cookies to `__Host-dtfs-session`

## Miscellaneous
* *Major* Cypress version update to `12.5.1`, this has introduced various breaking changes.
* `cy.session` being used for preserving cookies (session cookies used for login).
* Introduced custom cypress command `saveSession` responsible for saving `dtfs-session` and `csrf` cookies.
* Fixed various E2E tests effected by major version update.

### CODE REVIEW

File: e2e-tests/gef/cypress/e2e/about-exporter.spec.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ context('About Exporter Page', () => {
   });
 
   beforeEach(() => {
-    Cypress.Cookies.preserveOnce('connect.sid');
+    Cypress.Cookies.preserveOnce('dtfs-session');
```
Comment (IgnatG) [inline e2e-tests/gef/cypress/e2e/about-exporter.spec.js:42]: `Cypress.Cookies.preserveOnce('');` is deprecated in the latest version. If we need to update this line, then we may as well upgrade Cypress to its latest version and update the way we handle sessions in e2e tests --> https://docs.cypress.io/guides/references/changelog#12-0-0

**Review by IgnatG — CHANGES_REQUESTED** [review]
Upgrade Cypress to its latest version and change the `preserveOnce` functionality with `cy.session`

**Review by IgnatG — APPROVED** [review]
LGTM

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js [hunk 2]
```diff
@@ -44,34 +61,27 @@ context('Dashboard Deals filters - filter by multiple fields with multiple value
...
-    // apply filter 3
-    dashboardDeals.filters.panel.form.submissionType.MIA.checkbox().click();
-
-    // apply filter 4
-    dashboardDeals.filters.panel.form.submissionType.AIN.checkbox().click();
-
-    // submit filters
-    filters.panel.form.applyFiltersButton().click();
-
+  it('Should visit correct page', () => {
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js:74]: Not a blocker and very minor - seems the naming convention is lowercase

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-created-by-you.spec.js [hunk 3]
```diff
@@ -61,16 +61,22 @@ context('Dashboard Deals filters - Created by you', () => {
...
+  });
+
+  beforeEach(() => {
+    cy.saveSession();
+
     dashboardDeals.visit();
+    filters.showHideButton().click();
+  });
+
+  it('Visit the correct page', () => {
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-created-by-you.spec.js:73]: Not a blocker, could keep in line with the other tests `it('should redirect to the correct page'`

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js [hunk 4]
```diff
@@ -59,41 +59,62 @@ context('Dashboard Facilities filters - filter by product/facility type', () =>
         });
       });
     });
+
+    cy.login(BANK1_MAKER1);
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js:63]: Could this be in the `beforeEach` below?

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js [hunk 5]
```diff
@@ -105,32 +126,52 @@ context('Dashboard Facilities filters - filter by product/facility type', () =>
...
       shouldRenderAppliedFilterInMainContainerSelectedFilters(
         dashboardFacilities.filters.mainContainer.selectedFilters.typeContingent(),
         CONSTANTS.FACILITY.FACILITY_TYPE.CONTINGENT,
       );
     });
 
     it('renders only Contingent facilities', () => {
+      submitRedirectsToDashboard(dashboardFacilities.filters.panel.form.type.contingent.checkbox());
+
+      filters.showHideButton().click();
```
Comment (ttbarnes) [inline e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js:168]: These 2 lines are duplicated a lot in this file, is it possible to have in the `beforeEach` or is this required due to cypress upgrade?

File: gef-ui/server/middleware/headers/security.middleware.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ const security = (req, res, next) => {
   res.setHeader('X-Frame-Options', 'deny');
   res.setHeader('X-XSS-Protection', '1; mode=block');
   res.setHeader('X-Content-Type-Options', 'nosniff');
-  res.setHeader('Content-Security-Policy', "default-src 'none';connect-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self';form-action 'self';frame-ancestors 'self';img-src 'self';object-src 'none';script-src 'self' 'unsafe-inline';script-src-attr 'self' 'unsafe-inline';style-src 'self';upgrade-insecure-requests");
+  res.setHeader('Content-Security-Policy', "default-src 'none';connect-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' data:;form-action 'self';frame-ancestors 'self';img-src 'self';object-src 'none';script-src 'self' 'unsafe-inline';script-src-attr 'self' 'unsafe-inline';style-src 'self';upgrade-insecure-requests");
```
Comment (ttbarnes) [inline gef-ui/server/middleware/headers/security.middleware.js:26]: I don't remember the exact scripts setup with Portal and GEF, but I would double check in a dev/non-local environment that any inline scripts still work, e.g I think there is/was some inline scripts for Portal facility pages and maybe the main BSS/EWCS deal page.

In EXIP we discovered an issue after this same change in dev environment due to inline script not being executed, that is (I believe correctly) not picked up in E2E tests on a local machine or GHA environment.

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js (hunk 2 above)
Comment (ttbarnes) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js:74]: Few instances of this

**Review by ttbarnes — APPROVED** [review]

File: gef-ui/server/middleware/headers/security.middleware.js (hunk 6 above)
Comment (abhi-markan) [reply gef-ui/server/middleware/headers/security.middleware.js:26]: No `unsafe-inline` has been removed for DTFS. Thanks for flagging it up.

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js (hunk 2 above)
Comment (abhi-markan) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-by-multiple-fields-multiple-values.spec.js:74]: Renamed.

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-created-by-you.spec.js (hunk 3 above)
Comment (abhi-markan) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-deals/dashboard-deals-filter-created-by-you.spec.js:73]: Renamed.

File: e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js (hunk 5 above)
Comment (abhi-markan) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js:168]: They are, unfortunately not all tests cases are calling it.

Comment (abhi-markan) [reply e2e-tests/portal/cypress/e2e/journeys/dashboard/dashboard-facilities/dashboard-facilities-filter-by-product.spec.js:168]: `filters.showHideButton().click();` appears in beforeEach()

**Review by ttbarnes — APPROVED** [review]

**Review by Zainzzkk — APPROVED** [review]
Lgtm

**Pull request merged**